File: app/models/quizzes/quiz_question/answer_serializers/short_answer.rb
# frozen_string_literal: true

#
# Copyright (C) 2013 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module Quizzes::QuizQuestion::AnswerSerializers
  class ShortAnswer < Quizzes::QuizQuestion::AnswerSerializers::AnswerSerializer

    # Serialize a written, textual answer.
    #
    # Serialization request will be rejected if the answer isn't a string or is
    # too long. See Util#text_too_long?
    #
    # @param answer_hash[:text] String
    #   The textual/HTML answer. Will be text-escaped.
    #
    # @example output for an answer for QuizQuestion#1
    #  {
    #    :question_1 => "sanitized_answer"
    #  }
    def serialize(answer_text)
      rc = SerializedAnswer.new

      if !answer_text.is_a?(String)
        return rc.reject :invalid_type, 'answer', String
      elsif Util.text_too_long? answer_text
        return rc.reject :text_too_long
      end

      rc.answer[question_key] = Util.sanitize_text(answer_text)
      rc
    end

    # @return [String|NilClass] The textual answer, if any.
    def deserialize(submission_data, full=false)
      text = submission_data[question_key]

      if text.nil? || text.empty?
        return nil
      else
        text
      end
    end
  end
end

File: client_apps/canvas_quizzes/apps/statistics/js/views/questions/short_answer.js
/** @jsx React.DOM */
/*
 * Copyright (C) 2014 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

define(function(require) {
  var React = require('../../ext/react')
  var Question = require('jsx!../question')
  var CorrectAnswerDonut = require('jsx!../charts/correct_answer_donut')
  var DiscriminationIndex = require('jsx!../charts/discrimination_index')
  var AnswerTable = require('jsx!./answer_table')
  var calculateResponseRatio = require('../../models/ratio_calculator')
  var QuestionHeader = require('jsx!./header')

  var ShortAnswer = React.createClass({
    render: function() {
      var props = this.props
      var crr = calculateResponseRatio(props.answers, props.participantCount, {
        correctResponseCount: props.correct,
        questionType: props.questionType
      })

      return (
        <Question>
          <div className="grid-row">
            <div className="col-sm-8 question-top-left">
              <QuestionHeader
                responseCount={this.props.responses}
                participantCount={this.props.participantCount}
                questionText={this.props.questionText}
                position={this.props.position}
              />

              <div
                className="question-text"
                aria-hidden
                dangerouslySetInnerHTML={{__html: this.props.questionText}}
              />
            </div>
            <div className="col-sm-4 question-top-right" />
          </div>
          <div className="grid-row">
            <div className="col-sm-8 question-bottom-left">
              <AnswerTable answers={this.props.answers} />
            </div>
            <div className="col-sm-4 question-bottom-right">
              <CorrectAnswerDonut correctResponseRatio={crr} />
            </div>
          </div>
        </Question>
      )
    }
  })

  return ShortAnswer
})
